### ISSUE TITLE
Does YOLO V3 automatically perform data augmentation?

### ISSUE THREAD

**Issue by 89douner** [AlexeyAB/darknet#1815]
In YOLO V3 paper, they use multi-scale training, lots of data augmentation.

I don't know exactly whether they manually multiply original dataset through multi-scale training and data augmentation before training and use the multiplied dataset (original dataset+cropping and rotation dataset+scalining dataset).

As an additional question,
How does I check loss fluctuation through graph? In Tensorflow object detection code, they support tensorboard and I can check the loss graph. But, I can't check the graph in YOLO code. Please let me know!! Thanks !

**Comment by AlexeyAB** [comment]
* Yolo automatically do data augmentation during exactly training: https://github.com/AlexeyAB/darknet/blob/31ac46ba22112c8122d84a6a2b00db3a1ecb92cb/src/http_stream.cpp#L274-L335
    * randomly resize network, `random=1` in cfg-file
    * randomly crop part of image and randomly resize width and height, `jitter=0.3` in cfg-file
    * randomly change color (hue, saturation, exposure), `hue=0.1 saturation=1.5 exposure=1.5` in cfg-file

----

* If you will run compile this repository with `OPENCV=1`, then you can see Loss-graph during training

![chart_xnor_yolov3-tiny](https://user-images.githubusercontent.com/4096485/47426738-c1b91a80-d796-11e8-8ebd-34d4cdc710d7.jpg)

**Comment by 89douner** [comment]
Is there any method to see loss graph like above image in official version (pjreddie)? If you don't know, could I apply your code to the official version?

**Comment by AlexeyAB** [comment]
@89douner 

In original repo you can use another way.

* Run this command for training with logging:
`darknet.exe detector train data/voc.data cfg/yolov3-voc.cfg darknet53.conv.74 > test.log`
or to log and show console output:
`darknet.exe detector train data/voc.data cfg/yolov3-voc.cfg darknet53.conv.74 | tee test.log`

* After that training will be stopped, run this command to create loss-chart by using this script: https://github.com/AlexeyAB/darknet/tree/master/scripts/log_parser
`python log_parser.py --source-dir ./ --save-dir ./ --log-file test.log --show true`

**Comment by SpONA** [comment]
> * If you will run compile this repository with `OPENCV=1`, then you can see Loss-graph during training
> 
> ![chart_xnor_yolov3-tiny](https://user-images.githubusercontent.com/4096485/47426738-c1b91a80-d796-11e8-8ebd-34d4cdc710d7.jpg)

Sorry to bother you again, I got the loss graph, but the x-asix iteration number is 500200, so the points are concentrated to the very left of the graph when I only train like 1000 iterations. But in your graph, the x-axis of 10000 seems more suitable. Could you please tell me where should I change the code to set x-axis number?

**Comment by htt1084** [comment]
> * Yolo automatically do data augmentation during exactly training: 
>     
>       
>         [darknet/src/http_stream.cpp](https://github.com/AlexeyAB/darknet/blob/31ac46ba22112c8122d84a6a2b00db3a1ecb92cb/src/http_stream.cpp#L274-L335)
>       
>       
>           Lines 274 to 335
>         in
>         [31ac46b](/AlexeyAB/darknet/commit/31ac46ba22112c8122d84a6a2b00db3a1ecb92cb)
>       
>       
>       
>       
>   
>           
>             
>              image image_data_augmentation(IplImage* ipl, int w, int h, 
>           
>   
>           
>             
>                  int pleft, int ptop, int swidth, int sheight, int flip, 
>           
>   
>           
>             
>                  float jitter, float dhue, float dsat, float dexp) 
>           
>   
>           
>             
>              { 
>           
>   
>           
>             
>                  cv::Mat img = cv::cvarrToMat(ipl); 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                  // crop 
>           
>   
>           
>             
>                  cv::Rect src_rect(pleft, ptop, swidth, sheight); 
>           
>   
>           
>             
>                  cv::Rect img_rect(cv::Point2i(0, 0), img.size()); 
>           
>   
>           
>             
>                  cv::Rect new_src_rect = src_rect & img_rect; 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                  cv::Rect dst_rect(cv::Point2i(std::max<int>(0, -pleft), std::max<int>(0, -ptop)), new_src_rect.size()); 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                  cv::Mat cropped(cv::Size(src_rect.width, src_rect.height), img.type()); 
>           
>   
>           
>             
>                  cropped.setTo(cv::Scalar::all(0)); 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                  img(new_src_rect).copyTo(cropped(dst_rect)); 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                  // resize 
>           
>   
>           
>             
>                  cv::Mat sized; 
>           
>   
>           
>             
>                  cv::resize(cropped, sized, cv::Size(w, h), 0, 0, INTER_LINEAR); 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                  // flip 
>           
>   
>           
>             
>                  if (flip) { 
>           
>   
>           
>             
>                      cv::flip(sized, cropped, 1);    // 0 - x-axis, 1 - y-axis, -1 - both axes (x & y) 
>           
>   
>           
>             
>                      sized = cropped.clone(); 
>           
>   
>           
>             
>                  } 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                  // HSV augmentation 
>           
>   
>           
>             
>                  // CV_BGR2HSV, CV_RGB2HSV, CV_HSV2BGR, CV_HSV2RGB 
>           
>   
>           
>             
>                  if (ipl->nChannels >= 3) 
>           
>   
>           
>             
>                  { 
>           
>   
>           
>             
>                      cv::Mat hsv_src; 
>           
>   
>           
>             
>                      cvtColor(sized, hsv_src, CV_BGR2HSV);    // also BGR -> RGB 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                      std::vector<cv::Mat> hsv; 
>           
>   
>           
>             
>                      cv::split(hsv_src, hsv); 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                      hsv[1] *= dsat; 
>           
>   
>           
>             
>                      hsv[2] *= dexp; 
>           
>   
>           
>             
>                      hsv[0] += 179 * dhue; 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                      cv::merge(hsv, hsv_src); 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                      cvtColor(hsv_src, sized, CV_HSV2RGB);    // now RGB instead of BGR 
>           
>   
>           
>             
>                  } 
>           
>   
>           
>             
>                  else 
>           
>   
>           
>             
>                  { 
>           
>   
>           
>             
>                      sized *= dexp; 
>           
>   
>           
>             
>                  } 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                  //std::stringstream window_name; 
>           
>   
>           
>             
>                  //window_name << "augmentation - " << ipl; 
>           
>   
>           
>             
>                  //cv::imshow(window_name.str(), sized); 
>           
>   
>           
>             
>                  //cv::waitKey(0); 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                  // Mat -> IplImage -> image 
>           
>   
>           
>             
>                  IplImage src = sized; 
>           
>   
>           
>             
>                  image out = ipl_to_image(&src); 
>           
>   
>           
>             
>               
>           
>   
>           
>             
>                  return out; 
>           
>   
>           
>             
>              } 
>           
>       
>     
>   
>   
>   
>   * randomly resize network, `random=1` in cfg-file
>   * randomly crop part of image and randomly resize width and height, `jitter=0.3` in cfg-file
>   * randomly change color (hue, saturation, exposure), `hue=0.1 saturation=1.5 exposure=1.5` in cfg-file
> 
> * If you will run compile this repository with `OPENCV=1`, then you can see Loss-graph during training
> 
> ![chart_xnor_yolov3-tiny](https://user-images.githubusercontent.com/4096485/47426738-c1b91a80-d796-11e8-8ebd-34d4cdc710d7.jpg)

Hi Alexey, by checking the augmentation method, I'm not sure if jitter is even used in there?
Thank you!

**Comment by htt1084** [comment]
>   * randomly resize network, `random=1` in cfg-file
>   * randomly crop part of image and randomly resize width and height, `jitter=0.3` in cfg-file
>   * randomly change color (hue, saturation, exposure), `hue=0.1 saturation=1.5 exposure=1.5` in cfg-file

Hi Alexey, I've checked the above augmentation method, and I'm not sure if jitter is used in there?
Thank you!

**Comment by AlexeyAB** [comment]
@htt1084 

> Hi Alexey, I've checked the above augmentation method, and I'm not sure if jitter is used in there?
> Thank you!

What do you mean?

Just compile the latest version of Darknet and run training command with flag `-show_imgs`

**Comment by htt1084** [comment]
Hi Alexey, I meant that in the function below, "float jitter" is declared in the method, but it's never used inside this method itself. So I am a little confused about this.
> image image_data_augmentation(IplImage* ipl, int w, int h, 
>      int pleft, int ptop, int swidth, int sheight, int flip, 
>      float jitter, float dhue, float dsat, float dexp) 

On the other hand, I learn that jitter is used in here right? 
https://github.com/AlexeyAB/darknet/blob/master/src/data.c#L587
